### PR TITLE
MBS-7890: Fix reminder job for all DBMS

### PR DESCRIPTION
**Pull request by sh-csg**

### CODE REVIEW

File: classes/task/reminder.php [hunk 1]
```diff
@@ -52,16 +52,17 @@ public function get_name() {
     public function execute() {
         global $DB;
         $kanbancards = $DB->get_records_sql(
-            'SELECT ' . $DB->sql_concat('c.id', '"-"', 'a.userid') . ' as uniqid,
+            'SELECT ' . $DB->sql_concat('c.id', '\'-\'', 'a.userid') . ' as uniqid,
```
Comment (PhMemmel) [inline classes/task/reminder.php:55]: Everywhere else in the moodle core code similar issues are being solved by using
```suggestion
            'SELECT ' . $DB->sql_concat('c.id', "'-'", 'a.userid') . ' as uniqid,
```
So I suggest to use this solution.

BTW: I tried to do some research and it looks like (I could not really find a reliable source for this information) according to the ANSI standard escaping should be done by using double single quotes. Using a backslash only seems to be supported by MySQL and postgres, but, for example, not on Oracle (this information might be wrong or outdated). But let's go with the solution mentioned above anyway :-)

File: classes/task/reminder.php [hunk 2]
```diff
@@ -52,16 +52,17 @@ public function get_name() {
...
          INNER JOIN {kanban_assignee} a ON a.kanban_card = c.id
                 AND c.duedate != 0
                 AND c.reminder_sent = 0
                 AND c.completed = 0
-                AND (c.duedate < now() OR (c.reminderdate != 0 AND c.reminderdate < now()))
+                AND (c.duedate < :time OR (c.reminderdate != 0 AND c.reminderdate < :time2))
          INNER JOIN {kanban_board} b ON b.id = c.kanban_board
-         INNER JOIN {kanban} k ON b.kanban_instance = k.id'
+         INNER JOIN {kanban} k ON b.kanban_instance = k.id',
+            ['time' => time(), 'time2' => time()]
```
Comment (PhMemmel) [inline classes/task/reminder.php:65]: I suggest using a variable `$time = time();` and inject its value into the statement. If these two time calls differ very slightly and this statement ever shows up in an analysis tool one might wonder what these different times exactly mean :-)

**Review by PhMemmel — COMMENTED** [review]
If you are motivated enough, maybe this would also be a nice opportunity to add a unit test for this job. This would make sure there are no other incompatibilities.

File: classes/task/reminder.php (hunk 1 above)
Comment (sh-csg) [reply classes/task/reminder.php:55]: When using '\'', escaping is done by PHP so there is no SQL issue - but nevertheless, I'm choosing the "Moodle standard" ;-)

File: tests/reminder_test.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace mod_kanban;
```
Comment (PhMemmel) [inline tests/reminder_test.php:17]: According to coding style conventions, tests should match the namespace of the tested class as much as possible.

In this case this probably would be `mod_kanban\task`

Also there are some minor issues with codechecker. Could you please fix?

File: tests/reminder_test.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        $reminder->execute();
+
+        // User 0 and 2 should have one additional notification.
+
+        for ($i = 0; $i < 3; $i++) {
+            $this->assertEquals(
+                $notificationcount[$i] + (int)($i % 2 == 0),
+                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
+            );
+        }
```
Comment (PhMemmel) [inline tests/reminder_test.php:137]: I like line 134 very much, but it's really horrible to read :-D

```suggestion
        for ($i = 0; $i < 3; $i++) {
            // Zero notifications expected for user 1, one notification each expected for the other two users.
            $expectednotificationcount = $i === 1 ? 0 : 1;
            $this->assertEquals(
                $expectednotificationcount,
                $DB->count_records('notifications', ['useridfrom' => $noreply->id, 'useridto' => $this->users[$i]->id])
            );
        }
```

**Review by PhMemmel — COMMENTED** [review]
Thanks for adding the unit test. It only tests, if there are notifications are being sent, but I think that is ok for now.

File: classes/task/reminder.php (hunk 1 above)
Comment (PhMemmel) [reply classes/task/reminder.php]: ah, you're right of course. But yes, moodle "standard" :-)

File: tests/reminder_test.php (hunk 4 above)
Comment (PhMemmel) [reply tests/reminder_test.php]: Your solution is fine as well :+1:

**Pull request merged**